test/types.cc

```diff
@@ -454,3 +454,259 @@ copy_number: [100, 50, 1]
+        Status s;
+
+        for (const char* buf : good_examples) {
+            YAML::Node node = YAML::Load(buf);
+            unique_ptr<retained_format_field> rff;
+            s = retained_format_field::of_yaml(node, rff);
+            REQUIRE(s.ok());
+
+            YAML::Emitter ans;
+            s = rff->yaml(ans);
+            REQUIRE(s.ok());
+            node = YAML::Load(ans.c_str());
+            unique_ptr<retained_format_field> rff2;
+            s = retained_format_field::of_yaml(node, rff2);
+            REQUIRE(*rff == *rff2);
+        }
+    }
+
+    const char* buf4 = 1 + R"(
+        {orig_names: [XXX, YYY],
+         name: AAA,
+         description: foobar,
+         type: int,
+         number: GENOTYPE,
+         default_to_zero: false,
+         count: 5,
+         combi_method: zzzzz}
+)";
+    const char* bad_examples[] = {buf4};
+
+    SECTION("bad examples") {
+        Status s;
+
+        for (const char* buf : bad_examples) {
+            YAML::Node node = YAML::Load(buf);
+            unique_ptr<retained_format_field> rff;
+            s = retained_format_field::of_yaml(node, rff);
+            REQUIRE(s.bad());
+        }
+    }
+}
+
+
+TEST_CASE("genotyper_config") {
+    const char* buf1 = 1 + R"(
+         required_dp: 3
+         allele_dp_format: AD
+         ref_symbolic_allele: <NON_REF>
+         ref_dp_format: MIN_DP
+         output_residuals: true
+         output_format: VCF
+)";
+
+    const char* buf2 = 1 + R"(
+         required_dp: 0
+         allele_dp_format: AD
+         ref_symbolic_allele: <NON_REF>
+         ref_dp_format: MIN_DP
+         output_residuals: false
+         output_format: BCF
+)";
+
+     const char* buf3 = 1 + R"(
+required_dp: 0
+allele_dp_format: AD
+ref_symbolic_allele: <NON_REF>
+ref_dp_format: MIN_DP
+output_residuals: false
+output_format: BCF
+liftover_fields:
+
+  - name: AAA
+    description: foobar
+    type: int
+    number: genotype
+    default_to_zero: false
+    count: 5
+    combi_method: max
+    orig_names: [XXX, YYY]
+
+  - name: BBB
+    description: xxx
+    type: int
+    number: genotype
+    default_to_zero: false
+    count: 5
+    combi_method: max
+    orig_names: [uuu, zzz]
+
+ )";
+
+    const char* good_examples[] = {buf1, buf2, buf3};
+
+    SECTION("good examples") {
+        for (const char* buf : good_examples) {
+            Status s;
+            genotyper_config gc, gc2;
+            {
+                YAML::Node node = YAML::Load(buf);
+                s = genotyper_config::of_yaml(node, gc);
+                REQUIRE(s.ok());
+            }
+
+            {
+                YAML::Emitter ans;
+                s = gc.yaml(ans);
+                REQUIRE(ans.good());
+                REQUIRE(s.ok());
+
+                // **NOTE**
+                // I don't know why this is, but the Emitter
+                // adds quotation marks into the string, changing it,
+                // and preventing the YAML::Load step from succeeding.
```

Comment:
This seems disturbing. We'll need to be able to serialize and deserialize the configuration without this kind of manipulation. Can you look into it a little further?

interest: http://stackoverflow.com/a/22235064